File: Algo/Leetcode/215KthLargestNum.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        if(k<=0 || k>nums.size()) return -1;

        int start = 0; int end = nums.size()-1;
        int p = partition(nums, start, end);
        while(p != k-1){
            if (p>k-1)
                end = p-1;
            else
                start = p+1;
            p = partition(nums, start, end);
        }
        return nums[p];
    }

    int partition(vector<int>& nums, int start, int end){
        if(end<=start) return start;
        int pivot_idx = rand() % (end - start + 1) + start;
        swap(nums[pivot_idx], nums[end]);
        int idx = start - 1;
        for(int i=start; i<end; i++){
            if(nums[i] >= nums[end])
                swap(nums[++idx], nums[i]);
        }
        swap(nums[++idx], nums[end]);
        return idx;
    }
};

File: Algo/Leetcode/215KthLargestNum.py
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        n = len(nums)
        l, r = 0, n-1
        pivot = self.partition(nums, l, r)
        while pivot != k-1:
            if pivot < k-1:
                l = pivot + 1
            else:
                r = pivot - 1
            pivot = self.partition(nums, l, r)

        return nums[pivot]

    def partition(self, nums, l, r):
        if l >= r:
            return l

        target = nums[r]
        pivot = l
        for i in range(l, r):
            if nums[i] >= target:
                self.swap(nums, i, pivot)
                pivot += 1

        self.swap(nums, r, pivot)

        return pivot

    def swap(self, nums, i, j):
        tmp = nums[i]
        nums[i] = nums[j]
        nums[j] = tmp